[BONZO ALL OVER AGAIN!!!] Promote preact as default variant in discussion frontend

piuccio opened the pull request:
## What does this change?

Sorry for the caps.
After a long fiddling with (p)react and the nested dependencies, I've decided to use preact instead of react as our default library for discussion frontend.

I'm using `preact-compat` to maintain compatibility with react and `discussion-frontend` still generates both variants in case preact turns out to be sub-optimal in the future.

I managed to tree shake the following out of react
- Proptypes
- React.createClass
- Vendor prefixed `transition` and `animation` events
- React dev tools

but the result is:

React | Preact
----- | -----
48.2 KB | 15.6 KB

(* minified and gzipped)

`~32KB` bigger is a hard sell.

What's left inside React that doesn't seem very useful at the moment, but it's much harder to tree-shake is:

- Event normalization: I'm only using a click listener, but React adds support for touch events, transitions, scroll/wheel, audio/video playing events ...
- SVGs, they are XML so React can render them from code, fancy stuff but all my svgs are constant and injected as they are
- Refs to components, as in `<input ref={(input) => this._input = input} />`. I'm sure it's handy in some cases but I'm not using it and it takes space.

Preact doesn't even support them (well, maybe `ref`, I don't know)

__important__:
In order to reach all browsers, `discussion-frontend` allows to inject `Promise` as dependency, while it bundles `fetch` polyfill because it only contributes half a kilobyte gzipped.
The are better approaches but the reasoning is explained [here](https://github.com/guardian/discussion-frontend/pull/15).


## What is the value of this and can you measure success?

Remove a 100% audience AB test and use a single code base instead of the 3 options react, preact, old.

## Does this affect other platforms - Amp, Apps, etc?

No

<!--
Run the AMP test suite with `make validate-amp`

You should also validate a specific page that your change affects by adding the amp query string along with the development hash: http://localhost:3000/sport/2016/aug/25/katie-ledecky-first-pitch-washington-nationals-bryce-harper?amp=1#development=1

The AMP validation results will appear in your console.
-->

## Screenshots

## Request for comment

@guardian/guardian-frontend 
FYI @developit

<!--
*Does this PR meet the [contributing guidelines](https://github.com/guardian/frontend/blob/issue_pr_templates/.github/CONTRIBUTING.md#submission)?*
-->

dominickendrick commented on static/src/javascripts/projects/common/modules/discussion/discussion-frontend.js:
```diff
@@ -1,33 +1,39 @@
+
+        return require('discussion-frontend-' + VARIANT, function (frontend) {
             // Preact works in a slightly different way
             // https://github.com/developit/preact-compat/issues/145
```
This issue seems to be closed, does this code still need to be here ?

piuccio replied:
I didn't realize it was fixed, let me check.

dominickendrick commented on static/src/javascripts/projects/common/modules/discussion/discussion-frontend.js:
```diff
@@ -1,33 +1,39 @@
-        return (ab.isInVariant('DiscussionExternalFrontendAvatar', 'react') && curlConfig.paths['discussion-frontend-react']) ||
-            (ab.isInVariant('DiscussionExternalFrontendAvatar', 'preact') && curlConfig.paths['discussion-frontend-preact']);
-    }
+    var VARIANT = 'preact';
```
Is it worth setting the constant as we've decided to go with `preact` ? 
Why not just `require('discussion-frontend-preact')`,

piuccio replied:
just in case we want to use `'react'` in the future or you want to override it locally

developit commented:
Nice summary @piuccio! Just a note - inline svg and refs are both supported in Preact :tada:

piuccio commented:
@developit good to know, maybe I should try and tree-shake them 😁

piuccio commented:
@dominickendrick I've changed the code with your suggestions and removed some more dead code. Especially the template part, it now lives in `discussion-frontend`

desbo commented:
awesome work fabio

piuccio commented:
@desbo yet nobody gave a +1 😢

dominickendrick commented:
👍
